kv: don't send NoopRequest

tbg opened the pull request:
Also fixes funky code introduced ages ago<sup>[1]</sup> which meant that during
limited batches, results could remain unpopulated. We've never hit this in
practice, but it looked pretty bad.

Closes #16744.
Fixes #12142.

[1]:
https://github.com/cockroachdb/cockroach/pull/5092/files#diff-c7b3757cf7a8d326bfaf580795bad34dR775

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16946)
<!-- Reviewable:end -->

tschottdorf commented on pkg/kv/batch.go:
```diff
@@ -98,33 +97,32 @@ func truncate(ba roachpb.BatchRequest, rs roachpb.RSpan) (roachpb.BatchRequest,
 	}
 
-	var numNoop int
+	// TODO(tschottdorf): optimize so that we don't always make a new request
```
I'll probably address this before the merge.

tschottdorf commented:
BTW, this is basically a cleanup of https://github.com/cockroachdb/cockroach/pull/16744 with less code movement and some additional fixes.

tamird commented:
Various review comments from #16744 aren't addressed, right? Various mentions of NoopRequest remain in comments here.

---

Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3QArG_R8CttP0M7I:-Ko_3QArG_R8CttP0M7J:b-ch40vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L694)):*
> ```Go
> 
> 			if br == nil {
> 				// First response from a Range.
> ```

consider expanding this comment

---

*[pkg/kv/dist_sender.go, line 696 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3NURu9aiMQNQL0Sn:-Ko_2zR4U7X1SunWXzq7:byp4yq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L696)):*
> ```Go
> 				// First response from a Range.
> 				br = &roachpb.BatchResponse{}
> 				*br = *(resp.reply)
> ```

i don't think you need these parentheses

---

*[pkg/kv/dist_sender.go, line 701 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_40vIdHYuCAhn4dzO:-Ko_40vIdHYuCAhn4dzP:b-uqs8f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L701)):*
> ```Go
> 			// This was the second or later call in a cross-Range request.
> 			// Combine the new response with the existing one.
> 			if err := br.Combine(resp.reply, resp.positions); err != nil {
> ```

in the first response case you could just copy `resp.reply.Responses` above instead, right?

---

*[pkg/kv/dist_sender.go, line 1055 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_52fwXnh4rGmon34n:-Ko_52fwXnh4rGmon34o:bmlr702) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1055)):*
> ```Go
> 	// batch -- they'll all come back without any response since they never
> 	// execute.
> 	var scratchBA *roachpb.BatchRequest // allocate only when needed
> ```

I think this can be on the stack - you can still allocate the responses slice on demand, though.

---

*[pkg/kv/dist_sender.go, line 1057 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4OxJFsfDfS2eetg9:-Ko_4OxJFsfDfS2eetgA:b-v9lm7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1057)):*
> ```Go
> 	var scratchBA *roachpb.BatchRequest // allocate only when needed
> 	for i := range br.Responses {
> 		req := ba.Requests[i].GetInner()
> ```

seems like this can come after the early bail immediately below.

---

*[pkg/kv/dist_sender.go, line 1062 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4Y2kG3aXiOp5dBvr:-Ko_4Y2kG3aXiOp5dBvs:b-aydk9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1062)):*
> Not the most efficient
> 		// (but most convenient) way is to use

awkward sentence. perhaps "The most convenient (but not most efficient) way is..." would read better.

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-Ko_2XftIp2Ov4F-cnnI:bu0schx) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
> ```Go
> // TODO(tschottdorf): write tests.
> func (br *BatchResponse) Combine(otherBatch *BatchResponse, positions []int) error {
> 	for i, l := 0, len(otherBatch.Responses); i < l; i++ {
> ```

NYC but why do we need a binding for `l`?

---

*[pkg/roachpb/batch.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2i9lH9LyNb3ePzmV:-Ko_2i9lH9LyNb3ePzmW:bnkxa12) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L232)):*
> ```Go
> func (br *BatchResponse) Combine(otherBatch *BatchResponse, positions []int) error {
> 	for i, l := 0, len(otherBatch.Responses); i < l; i++ {
> 		id := positions[i]
> ```

s/id/pos/g?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Ko_1pYjXnvUKkjKxX2y:b-mu2niv)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Cleaned up all the NoopRequests mentions I could find. Note that the actual NoopRequest is still around. We'll have to migrate it out over a release that can only be migrated to from the previous version. Maybe the Raft log separate engine change needs that anyway?

---

Review status: 2 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3QArG_R8CttP0M7I:-Ko_74eHEHOtSZ3UmzYq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L694)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consider expanding this comment
</blockquote></details>

Done.

---

*[pkg/kv/dist_sender.go, line 696 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3NURu9aiMQNQL0Sn:-Ko_7BAe16cRLFfXh4Dg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L696)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i don't think you need these parentheses
</blockquote></details>

Done.

---

*[pkg/kv/dist_sender.go, line 701 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_40vIdHYuCAhn4dzO:-Ko_7Bz7pEmzxUWyOtyl:bs0t22w) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L701)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

in the first response case you could just copy `resp.reply.Responses` above instead, right?
</blockquote></details>

No, if the first request didn't have all of the batch active, the response slice is shorter and needs to be mapped (just like any other request). Example: `Put(range1), Put(range2)`.

---

*[pkg/kv/dist_sender.go, line 1055 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_52fwXnh4rGmon34n:-Ko_7cWB9cbXj9f2w5iN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1055)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think this can be on the stack - you can still allocate the responses slice on demand, though.
</blockquote></details>

Done.

---

*[pkg/kv/dist_sender.go, line 1057 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4OxJFsfDfS2eetg9:-Ko_7fEF2SMKqVMnZ0Lc:bmohmf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1057)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

seems like this can come after the early bail immediately below.
</blockquote></details>

Not trivially, `br` is not available down there and I also want to avoid making assumptions that just because there's a batch-wide limit, the execution will always be serial.

---

*[pkg/kv/dist_sender.go, line 1062 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4Y2kG3aXiOp5dBvr:-Ko_97zNCksiPiafTBom:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1062)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> Not the most efficient
> 		// (but most convenient) way is to use

awkward sentence. perhaps "The most convenient (but not most efficient) way is..." would read better.
</blockquote></details>

Done.

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-Ko_9Gxi1a2ey2dV4RTc:b-6hrz4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

NYC but why do we need a binding for `l`?
</blockquote></details>

To avoid calling `len` in a loop? Probably not a detectable difference, but not doing work in the predicate seems like a good pattern in general.

---

*[pkg/roachpb/batch.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2i9lH9LyNb3ePzmV:-Ko_9NcV7BJ3nqUtXRSp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L232)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/id/pos/g?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Ko_BYd7oD6BQyQRqi1q:bjyjagf)*
<!-- Sent from Reviewable.io -->

tamird commented:
You amended the wrong commit in addressing my comments.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/kv/batch.go, line 31 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_Dhh6URB4oPuP2yvR:-Ko_Dhh6URB4oPuP2yvS:b8ua2rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/71188884ed3cffbe95b646c5774a84445d20688b/pkg/kv/batch.go#L31)):*
> active key range

is that different from "the given span" mentioned in the previous sentence?

---

*[pkg/kv/dist_sender.go, line 694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3QArG_R8CttP0M7I:-Ko_D3Ldnw_kIkfJ8u9d:b-779wjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L694)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think it'd be useful to explain why we need to "bootstrap" this response with a shallow copy.

---

*[pkg/kv/dist_sender.go, line 1057 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4OxJFsfDfS2eetg9:-Ko_D7PzDUAl2kt3oV-H:b7atvjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1057)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not trivially, `br` is not available down there and I also want to avoid making assumptions that just because there's a batch-wide limit, the execution will always be serial.
</blockquote></details>

I don't understand; I'm just suggesting moving the call to `GetInner` below the `continue`.
```
		if br.Responses[i] != (roachpb.ResponseUnion{}) {
			continue
		}
		req := ba.Requests[i].GetInner()
```

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-Ko_CK1uWja4Gui-S62s:b-lzaau3) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

To avoid calling `len` in a loop? Probably not a detectable difference, but not doing work in the predicate seems like a good pattern in general.
</blockquote></details>

I must be missing something; ISTM that this loop can be written `for i := 0; i < len(otherBatch.Responses); i++ {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Ko_DuM_jeT6xaBCt6IZ:be2gfbg)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 1062 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KobrqUZEAQtw6Gl7phY:-KobrqUZEAQtw6Gl7phZ:bfxf5eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/71188884ed3cffbe95b646c5774a84445d20688b/pkg/kv/dist_sender.go#L1062)):*
> ```Go
> 		}
> 		// We need to summon an empty response. The most convenient (but not
> 		// most efficient) way is to use (*BatchRequest).CreateReply.
> ```

We could auto-generate `func (*xyzRequest) CreateReply() *xyzResponse` methods. See `roachpb/gen_batch.go`. Or perhaps a reorg of the existing `BatchRequest.CreateReply` to allow passing in the `BatchResponse` and have it only fill empty slots. Not needed for this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-KobshNx2yL_9P9bLISp:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/kv/batch.go, line 100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_0dD4QvVa8L6Afefb-r2-100:-Kocdb49h0M6DL-pxO8W:b22fonh) ([raw file](https://github.com/cockroachdb/cockroach/blob/36e1a9d0dd03a721bf73a24b3ae0dfc57938bdb7/pkg/kv/batch.go#L100)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'll probably address this before the merge.
</blockquote></details>

Ok, decided not to.

---

*[pkg/kv/batch.go, line 31 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_Dhh6URB4oPuP2yvR:-KoccRJIByLRJVuHsQXA:b-plpz83) ([raw file](https://github.com/cockroachdb/cockroach/blob/71188884ed3cffbe95b646c5774a84445d20688b/pkg/kv/batch.go#L31)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> active key range

is that different from "the given span" mentioned in the previous sentence?
</blockquote></details>

No, reverted to "the given span" which is clearer.

---

*[pkg/kv/dist_sender.go, line 694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_3QArG_R8CttP0M7I:-KocWfM0Teqv8Vfj3x97:b-ns4k96) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L694)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think it'd be useful to explain why we need to "bootstrap" this response with a shallow copy.
</blockquote></details>

Yikes, looking at this more, it's seriously funky code again. We're just throwing away all the other updates. I hope I didn't review any of this when it happened... Luckily we did update the Txn, which is the only thing that mattered for correctness. Definitely lost snowball traces for all but the first range touched, though.

---

*[pkg/kv/dist_sender.go, line 1057 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_4OxJFsfDfS2eetg9:-KocQWzYEFD870tRgiR1:bknw49k) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/kv/dist_sender.go#L1057)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't understand; I'm just suggesting moving the call to `GetInner` below the `continue`.
```
		if br.Responses[i] != (roachpb.ResponseUnion{}) {
			continue
		}
		req := ba.Requests[i].GetInner()
```
</blockquote></details>

Oh, sorry. Yes, absolutely. Done.

---

*[pkg/kv/dist_sender.go, line 1062 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KobrqUZEAQtw6Gl7phY:-KocQRUUUQlJscRIqARi:b5auydq) ([raw file](https://github.com/cockroachdb/cockroach/blob/71188884ed3cffbe95b646c5774a84445d20688b/pkg/kv/dist_sender.go#L1062)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We could auto-generate `func (*xyzRequest) CreateReply() *xyzResponse` methods. See `roachpb/gen_batch.go`. Or perhaps a reorg of the existing `BatchRequest.CreateReply` to allow passing in the `BatchResponse` and have it only fill empty slots. Not needed for this PR.
</blockquote></details>

Yeah, I saw that. Decided to punt. Added TODO, though.

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-KocPsxgWAD_RWbYu-DE:b-knw3bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I must be missing something; ISTM that this loop can be written `for i := 0; i < len(otherBatch.Responses); i++ {`
</blockquote></details>

that would call `len(...)` a linear number of times, right? I think that motivates the pattern. See https://play.golang.org/p/6og6Rw7t44

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@tamird commit history cleaned up.

---

Review status: 2 of 7 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Koce8UTGTo-NPPkdqw1:b36o9sg)*
<!-- Sent from Reviewable.io -->

tamird commented:
Heh, I'm glad I prodded you on those comments. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a couple small nits.

perhaps add a unit test for Combine, too?

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/kv/dist_sender.go, line 693 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KociQB7GQswc6L6A_Jm:-KociQB7GQswc6L6A_Jn:bltmacg) ([raw file](https://github.com/cockroachdb/cockroach/blob/787916c9a231909306e1f0f0bc28465cc3e856dc/pkg/kv/dist_sender.go#L693)):*
> , and thus our initial response.

this is no longer true. just remove this sentence?

also, can't we just remove this cleverness? allocate the response slice up front and be done with it.

better yet you could make the return a value rather than a pointer, but I'd understand if you wanted to punt on that.

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-Kocj2NsRQ4OIVfq0yFM:b-x3tmkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

that would call `len(...)` a linear number of times, right? I think that motivates the pattern. See https://play.golang.org/p/6og6Rw7t44
</blockquote></details>

Oh, well then you could just write `for i := range otherBatch.Responses`, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-KocjQ0HlPBW8bjR2GRp:b-1gotn5)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/batch.go, line 229 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KocjlakhrmNx2Myt_US:-KocjlakhrmNx2Myt_UT:bb6wze0) ([raw file](https://github.com/cockroachdb/cockroach/blob/787916c9a231909306e1f0f0bc28465cc3e856dc/pkg/roachpb/batch.go#L229)):*
> ```Go
> // of slots must be equal and the respective slots must be combinable.
> // On error, the receiver BatchResponse is in an invalid state.
> // TODO(tschottdorf): write tests.
> ```

Heh.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Added unit test. TFTRs!

---

Review status: 4 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 693 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KociQB7GQswc6L6A_Jm:-Koeqa1MSVJJHmt2VxCH:b-gzmj9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/787916c9a231909306e1f0f0bc28465cc3e856dc/pkg/kv/dist_sender.go#L693)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> , and thus our initial response.

this is no longer true. just remove this sentence?

also, can't we just remove this cleverness? allocate the response slice up front and be done with it.

better yet you could make the return a value rather than a pointer, but I'd understand if you wanted to punt on that.
</blockquote></details>

Done, though it's still a pointer. I'm not sure it's supposed to return a value, though the current lax discipline of returning a non-nil batch in the error case suggests perhaps. Feel free to pull this thread further, but I'll leave it at that.

---

*[pkg/roachpb/batch.go, line 231 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-Ko_2XftIp2Ov4F-cnnH:-Koer0551bZS-hmUpv9z:b-ze4896) ([raw file](https://github.com/cockroachdb/cockroach/blob/deedb53702789204cb7ea0ce7c6c78b2a821c6cd/pkg/roachpb/batch.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Oh, well then you could just write `for i := range otherBatch.Responses`, right?
</blockquote></details>

Heh, can't argue with that. Changed.

---

*[pkg/roachpb/batch.go, line 229 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KocjlakhrmNx2Myt_US:-Kof-Na2kTVNIoDJ3n4m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/787916c9a231909306e1f0f0bc28465cc3e856dc/pkg/roachpb/batch.go#L229)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Heh.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Kof-Prtw6XjYlLXY42i:bwvbr09)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Kog7uAovXJxMtlVIINa:b-uwkowd)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/kv/dist_sender.go, line 1050 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KoglT9blv4UntUpgmy3:-KoglT9blv4UntUpgmy4:b7hxc9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5db0ed61d0c4b9e46b709d1add18775b15bd767c/pkg/kv/dist_sender.go#L1050)):*
> ```Go
> ) {
> 	// Some requests might have no response at all if we used a batch-wide
> 	// limit; simply create trivial responses for those. Note that any type
> ```

Was the problem here that we were only handling a few specific request types? I believe that in practice we never combine limited scans with any other requests, which would explain why it was never a problem. Also, you said "results could remain unpopulated" - wouldn't we have hit the `panic` in this case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: don't send NoopRequest

Also fixes funky code introduced ages ago[1] which meant that during
limited batches, results could remain unpopulated. We've never hit
this in practice, but it looked pretty bad.

Closes #16744.
Fixes #12142.

[1]: https://github.com/cockroachdb/cockroach/pull/5092/files#diff-c7b3757cf7a8d326bfaf580795bad34dR775

tschottdorf commented:
One last small change to address a newly tickled data race + leaving a TODO... will merge on green. Astonishing that we overlooked the actual meat of this change for so long, such a straight-up win.

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 1050 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KoglT9blv4UntUpgmy3:-Koh11cr798BTSUtokkX:b-j9q8g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5db0ed61d0c4b9e46b709d1add18775b15bd767c/pkg/kv/dist_sender.go#L1050)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Was the problem here that we were only handling a few specific request types? I believe that in practice we never combine limited scans with any other requests, which would explain why it was never a problem. Also, you said "results could remain unpopulated" - wouldn't we have hit the `panic` in this case?
</blockquote></details>

Yep, that's it. As you mentioned we got away with it because of our limited use of this feature -- yes, we would've hit the panic, but that's pretty bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-:-Koh2Wc_E9ESYgie3x2u:b-kwsbmu)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/kv/dist_sender.go, line 1050 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946#-KoglT9blv4UntUpgmy3:-Koh88dyfANTftE4f9cP:b-9s1hqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5db0ed61d0c4b9e46b709d1add18775b15bd767c/pkg/kv/dist_sender.go#L1050)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yep, that's it. As you mentioned we got away with it because of our limited use of this feature -- yes, we would've hit the panic, but that's pretty bad.
</blockquote></details>

Cool, thanks. This part of the change LGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16946)*
<!-- Sent from Reviewable.io -->
